x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,90 +5,393 @@
+                params.getFollowShardId(), currentIndexMetadataVersion, minimumRequiredIndexMetadataVersion);
+            updateMapping(imdVersion -> {
+                retryCounter.set(0);
+                currentIndexMetadataVersion = imdVersion;
+                task.run();
+            });
+        }
+    }
+
+    void handleFailure(Exception e, Runnable task) {
+        assert e != null;
+        if (shouldRetry(e)) {
+            if (isStopped() == false && retryCounter.incrementAndGet() <= RETRY_LIMIT) {
+                LOGGER.debug(new ParameterizedMessage("{} error during follow shard task, retrying...", params.getFollowShardId()), e);
+                scheduler.accept(retryTimeout, task);
+            } else {
+                markAsFailed(new ElasticsearchException("retrying failed [" + retryCounter.get() +
+                    "] times, aborting...", e));
+            }
+        } else {
+            markAsFailed(e);
+        }
+    }
+
+    private boolean shouldRetry(Exception e) {
+        return NetworkExceptionHelper.isConnectException(e) ||
+            NetworkExceptionHelper.isCloseConnectionException(e) ||
+            TransportActions.isShardNotAvailableException(e);
+    }
+
+    // These methods are protected for testing purposes:
+    protected abstract void updateMapping(LongConsumer handler);
+
+    protected abstract void innerSendBulkShardOperationsRequest(List<Translog.Operation> operations, LongConsumer handler,
+                                                                Consumer<Exception> errorHandler);
+
+    protected abstract void innerSendShardChangesRequest(long from, int maxOperationCount, Consumer<ShardChangesAction.Response> handler,
+                                                         Consumer<Exception> errorHandler);
+
     @Override
     protected void onCancelled() {
         markAsCompleted();
     }
 
-    public boolean isRunning() {
-        return isCancelled() == false && isCompleted() == false;
-    }
-
-    void updateProcessedGlobalCheckpoint(long processedGlobalCheckpoint) {
-        this.processedGlobalCheckpoint.set(processedGlobalCheckpoint);
+    protected boolean isStopped() {
+        return isCancelled() || isCompleted();
     }
 
     @Override
-    public Task.Status getStatus() {
-        return new Status(processedGlobalCheckpoint.get());
+    public Status getStatus() {
+        return new Status(globalCheckpoint, lastRequestedSeqno, processedGlobalCheckpoint, numConcurrentReads, numConcurrentWrites);
     }
 
     public static class Status implements Task.Status {
 
         public static final String NAME = "shard-follow-node-task-status";
 
-        static final ParseField PROCESSED_GLOBAL_CHECKPOINT_FIELD = new ParseField("processed_global_checkpoint");
+        static final ParseField GLOBAL_CHECKPOINT_FIELD = new ParseField("leader_global_checkpoint");
```

Comment:
can we align the name with the field (LEADER_GLOBAL_CHECKPOINT_FIELD & FOLLOWER_GLOBAL_CHECKPOINT_FIELD)?